simtool.cc

```diff
@@ -7150,6 +7150,92 @@ const char *tool_stop_mover_t::do_work( player_t *player, const koord3d &last_po
 	return NULL;
 }
 
+bool tool_change_city_of_building_t::init(player_t* player) {
+	two_click_tool_t::init(player);
+
+	// convert default_param to city and update highlight_city
+	highlight_city = welt->get_cities()[atoi(default_param)];
+	for( int i = 0; i < welt->get_cities().get_count(); i++) {
+		city_info_t* w = dynamic_cast<city_info_t*>(win_get_top());
+		if( w  &&  w->get_city() == welt->get_cities()[i] ) {
+			w->update(highlight_city);
+		}
+	}
+	return true;
```

Comment:
The function should return false here to reject the initialization if `highlight_city` is not valid.